README.md
# banai_delayed_displays

You can delay the display of widgets with gradient and offset animations.
When your widgets are very performance-intensive and cause routing jumps to freeze, you can also pass in a placeholder to improve performance.

## How to cancel animation redrawing in listview

When using components, just pass in a unique key, for example:
```
const BanaiDelayedDisplays(
	key: Key('1'),  // add unique key
	fadeIn: true,
	from: Offset(0, 0.35),
	delay: Duration(milliseconds: 300),
	duration: Duration(milliseconds: 500),
	placelholder: Text(''),
	child: Center(
		child: Text('delayed_displays_example'),
	),
),
```



## Preview
![preview](https://raw.githubusercontent.com/ihongwu/banai_delayed_displays/main/documentation/preview.gif)


## Super simple to use
```

final BanaiDelayedDisplaysController _banaiDelayedDisplaysController = BanaiDelayedDisplaysController();

DelayedDisplays(
	banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
	from: const Offset(-0.35, 0.35),
	delay: const Duration(milliseconds: 300),
	duration: const Duration(milliseconds: 500),
	child: Container(
		width: 200,
		height: 50,
		color: Colors.blue,
		child: const Text('delayed_displays_example'),
	),
),
```

```
@override
void dispose() {
	super.dispose();
	_banaiDelayedDisplaysController.dispose();
}
```


## Getting Started
Add dependency
```
dependencies:
  delayed_displays: ^last version
```

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.



CHANGELOG.md
## 1.0.3

* update README.md

## 1.0.2

* update README.md

## 1.0.1

* Add BanaiDelayedDisplaysController.

## 1.0.0

* TODO: Describe initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:banai_delayed_displays_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/banai_delayed_displays_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('banai_delayed_displays');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });
}


lib/banai_delayed_displays.dart
import 'dart:async';
import 'package:flutter/material.dart';
// List keys = [];
class BanaiDelayedDisplays extends StatefulWidget {
  /// Child that will be displayed with the animation and delay
  final Widget child;

  /// The placeholder widget sets components that are as close to the actual widget as possible
  final Widget ? placelholder;

  /// Delay before displaying the widget and the animations
  final Duration delay;

  /// Duration of the fading animation
  final Duration duration;

  /// Curve of the sliding animation
  final Curve curve;

  /// Offset of the widget at the beginning of the sliding animation
  final Offset from;

  /// If true, make the child appear, disappear otherwise. Default to true.
  final bool fadeIn;

  // _banaiDelayedDisplaysController
  final BanaiDelayedDisplaysController banaiDelayedDisplaysController;
  const BanaiDelayedDisplays(
    {Key? key,
    required this.child, 
    required this.banaiDelayedDisplaysController,
    this.placelholder,
    this.delay = const Duration(milliseconds: 420), 
    this.duration = const Duration(milliseconds: 600), 
    this.curve = Curves.decelerate, 
    this.from = const Offset(0, 0),
    this.fadeIn = true}) : super(key: key);
  @override
  State<BanaiDelayedDisplays> createState() => _BanaiDelayedDisplaysState();
}

class _BanaiDelayedDisplaysState extends State<BanaiDelayedDisplays> with TickerProviderStateMixin{
  
  // whether to display child
  bool showChild = false;

  /// last displayed widget
  Widget ? result;

  /// Controller of the opacity animation
  late AnimationController _opacityController;

  /// Sliding Animation offset
  late Animation<Offset> _slideAnimationOffset;

  late Duration delay;

  /// Timer used to delayed animation
  Timer ? _timer;


  @override
  void initState() {
    super.initState();

    List keys = widget.banaiDelayedDisplaysController.getKeysList();

    delay = widget.delay;
    var key = widget.key;

    bool isanimate = true;
    
    if(keys.contains(key.toString())) {
      isanimate = false;
      delay = Duration.zero;
    }

    _opacityController = AnimationController(
      vsync: this,
      duration: isanimate ? widget.duration : Duration.zero,
    );

    final CurvedAnimation curvedAnimation = CurvedAnimation(
      curve: widget.curve,
      parent: _opacityController,
    );

    _slideAnimationOffset = Tween<Offset>(
      begin: widget.from,
      end: Offset.zero,
    ).animate(curvedAnimation);
    
    if(key != null) {
      if(!keys.contains(key.toString())) {
        keys.add(key.toString());
      }
    }
    
    _runFadeAnimation();
  }


  @override
  void didUpdateWidget(BanaiDelayedDisplays oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (
      oldWidget.fadeIn != widget.fadeIn
    ) {
      _runFadeAnimation();
    }

    if(
      oldWidget.child != widget.child ||
      oldWidget.curve != widget.curve || 
      oldWidget.delay != widget.delay || 
      oldWidget.duration != widget.duration || 
      oldWidget.from != widget.from || 
      oldWidget.placelholder != widget.placelholder
    ) {
      setState(() {});
    }
    return ;
  }


  void _runFadeAnimation() {
    _timer = Timer(delay, () {
      result = widget.child;
      showChild = true;
      setState(() {});
      widget.fadeIn ? _opacityController.forward() : _opacityController.reverse();
    });
  }

  /// Dispose the opacity controller
  @override
  void dispose() {
    _opacityController.dispose();
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return showChild || widget.placelholder == null ? FadeTransition(
        opacity: _opacityController,
        child: SlideTransition(
          position: _slideAnimationOffset,
          child: widget.child,
        ),
      ) : widget.placelholder!;
  }
}


class BanaiDelayedDisplaysController {
  List<String> keys = [];
  List<String> getKeysList(){
    return keys;
  }

  void addKeys(String key){
    keys.add(key);
  }

  void dispose(){
    keys.clear();
  }
}

android/src/main/java/cn/com/banai/banai_delayed_displays/BanaiDelayedDisplaysPlugin.java
package cn.com.banai.banai_delayed_displays;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** BanaiDelayedDisplaysPlugin */
public class BanaiDelayedDisplaysPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "banai_delayed_displays");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:banai_delayed_displays/banai_delayed_displays.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final BanaiDelayedDisplaysController _banaiDelayedDisplaysController = BanaiDelayedDisplaysController();
  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
    _banaiDelayedDisplaysController.dispose();
  }
  

  @override
  Widget build(BuildContext context) {
    return  MaterialApp(
      home: Scaffold(
        backgroundColor: Colors.pinkAccent,
        body: SizedBox(
          width: double.infinity,
          child: ListView(
            children:  [
              
              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                key: Key('1'),
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              GestureDetector(
                child: BanaiDelayedDisplays(
                  banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                  key: const Key('2'),
                  fadeIn: true,
                  from: const Offset(0, 0.35),
                  delay: const Duration(milliseconds: 600),
                  duration: const Duration(milliseconds: 500),
                  placelholder: Center(
                    child: Container(
                      width: 200,
                      height: 50,
                      color: Colors.transparent,
                    ),
                  ),
                  child: Center(
                    child: Container(
                      width: 200,
                      height: 50,
                      color: Colors.blue,
                      child: const Center(
                        child: Text('delayed_displays_example'),
                      ),
                    ),
                  ),
                ),
                onTap: (){
                  _banaiDelayedDisplaysController.dispose();
                },
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                key: Key('3'),
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                key: const Key('4'),
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                key: Key('6'),
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                key: const Key('7'),
                fadeIn: true,
                from: const Offset(0, 0.35),
                delay: const Duration(milliseconds: 600),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Center(
                      child: Text('delayed_displays_example'),
                    ),
                  ),
                ),
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(0, 0.35),
                delay: const Duration(milliseconds: 600),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Center(
                      child: Text('delayed_displays_example'),
                    ),
                  ),
                ),
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(0, 0.35),
                delay: const Duration(milliseconds: 600),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Center(
                      child: Text('delayed_displays_example'),
                    ),
                  ),
                ),
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(0, 0.35),
                delay: const Duration(milliseconds: 600),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Center(
                      child: Text('delayed_displays_example'),
                    ),
                  ),
                ),
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(0, 0.35),
                delay: const Duration(milliseconds: 600),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Center(
                      child: Text('delayed_displays_example'),
                    ),
                  ),
                ),
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(0, 0.35),
                delay: const Duration(milliseconds: 600),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Center(
                      child: Text('delayed_displays_example'),
                    ),
                  ),
                ),
              ),


              Container(
                height: 100,
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: Offset(0.35, 0.35),
                delay: Duration(milliseconds: 300),
                duration: Duration(milliseconds: 500),
                placelholder: Text(''),
                child: Center(
                  child: Text('delayed_displays_example'),
                ),
              ),

              BanaiDelayedDisplays(
                banaiDelayedDisplaysController: _banaiDelayedDisplaysController,
                fadeIn: true,
                from: const Offset(-0.35, 0.35),
                delay: const Duration(milliseconds: 300),
                duration: const Duration(milliseconds: 500),
                placelholder: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.transparent,
                  ),
                ),
                child: Center(
                  child: Container(
                    width: 200,
                    height: 50,
                    color: Colors.blue,
                    child: const Text('delayed_displays_example'),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
